File: src-tauri/src/markdown/frontmatter.rs
use crate::db::models::Note;
use yaml_rust2::{Yaml, YamlLoader};

#[derive(Debug, Clone)]
pub struct FrontmatterData {
    pub id: Option<String>,
    pub title: Option<String>,
    pub tags: Vec<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub is_pinned: bool,
}

/// Serialize a Note into a markdown string with YAML frontmatter.
pub fn serialize_frontmatter(note: &Note) -> String {
    let mut fm = String::new();
    fm.push_str("---\n");
    fm.push_str(&format!("id: \"{}\"\n", note.id));
    fm.push_str(&format!("title: \"{}\"\n", note.title.replace('"', "\\\"")));
    if !note.tags.is_empty() {
        fm.push_str("tags:\n");
        for tag in &note.tags {
            fm.push_str(&format!("  - \"{}\"\n", tag));
        }
    } else {
        fm.push_str("tags: []\n");
    }
    fm.push_str(&format!("created_at: \"{}\"\n", note.created_at));
    fm.push_str(&format!("updated_at: \"{}\"\n", note.updated_at));
    fm.push_str(&format!("is_pinned: {}\n", note.is_pinned));
    fm.push_str("---\n");
    fm.push_str(&note.content);
    fm
}

/// Parse a markdown string with YAML frontmatter into FrontmatterData and body content.
pub fn parse_frontmatter(content: &str) -> Result<(FrontmatterData, String), String> {
    if !content.starts_with("---") {
        return Ok((
            FrontmatterData {
                id: None,
                title: None,
                tags: vec![],
                created_at: None,
                updated_at: None,
                is_pinned: false,
            },
            content.to_string(),
        ));
    }

    let rest = &content[3..];
    let end = rest
        .find("\n---\n")
        .or_else(|| rest.find("\n---"))
        .ok_or_else(|| "Invalid frontmatter: no closing ---".to_string())?;

    let yaml_str = &rest[..end];
    let body_start = end + 4; // skip \n---\n
    let body = if body_start < rest.len() {
        rest[body_start..].to_string()
    } else {
        String::new()
    };

    let docs = YamlLoader::load_from_str(yaml_str.trim())
        .map_err(|e| format!("YAML parse error: {}", e))?;

    if docs.is_empty() {
        return Ok((
            FrontmatterData {
                id: None,
                title: None,
                tags: vec![],
                created_at: None,
                updated_at: None,
                is_pinned: false,
            },
            body,
        ));
    }

    let doc = &docs[0];

    let id = doc["id"].as_str().map(String::from);
    let title = doc["title"].as_str().map(String::from);
    let created_at = doc["created_at"].as_str().map(String::from);
    let updated_at = doc["updated_at"].as_str().map(String::from);
    let is_pinned = doc["is_pinned"].as_bool().unwrap_or(false);

    let tags = match &doc["tags"] {
        Yaml::Array(arr) => arr
            .iter()
            .filter_map(|t| t.as_str().map(String::from))
            .collect(),
        _ => vec![],
    };

    Ok((
        FrontmatterData {
            id,
            title,
            tags,
            created_at,
            updated_at,
            is_pinned,
        },
        body,
    ))
}

File: mobile/src/services/frontmatter.ts
import type { FrontmatterData } from "@/types";

interface SerializableNote {
  id: string;
  title: string;
  content: string;
  tags: string[];
  created_at: string;
  updated_at: string;
  is_pinned: boolean;
}

/**
 * Serialize a Note into a markdown string with YAML frontmatter.
 * Port of Rust serialize_frontmatter — uses manual string building for exact format match.
 */
export function serializeFrontmatter(note: SerializableNote): string {
  let fm = "";
  fm += "---\n";
  fm += `id: "${note.id}"\n`;
  fm += `title: "${note.title.replace(/"/g, '\\"')}"\n`;
  if (note.tags.length > 0) {
    fm += "tags:\n";
    for (const tag of note.tags) {
      fm += `  - "${tag}"\n`;
    }
  } else {
    fm += "tags: []\n";
  }
  fm += `created_at: "${note.created_at}"\n`;
  fm += `updated_at: "${note.updated_at}"\n`;
  fm += `is_pinned: ${note.is_pinned}\n`;
  fm += "---\n";
  fm += note.content;
  return fm;
}

/**
 * Parse a markdown string with YAML frontmatter into FrontmatterData and body content.
 * Port of Rust parse_frontmatter — manual YAML parsing for exact compatibility.
 */
export function parseFrontmatter(content: string): {
  frontmatter: FrontmatterData;
  body: string;
} {
  const emptyResult: FrontmatterData = {
    id: null,
    title: null,
    tags: [],
    created_at: null,
    updated_at: null,
    is_pinned: false,
  };

  if (!content.startsWith("---")) {
    return { frontmatter: emptyResult, body: content };
  }

  const rest = content.slice(3);

  // Find closing --- delimiter. Try \n---\n first (has body after), then \n--- at end.
  let yamlEnd: number;
  let bodyOffset: number;

  const posWithBody = rest.indexOf("\n---\n");
  if (posWithBody !== -1) {
    yamlEnd = posWithBody;
    bodyOffset = posWithBody + 5; // skip \n---\n
  } else {
    const posAtEnd = rest.indexOf("\n---");
    if (posAtEnd !== -1) {
      const after = posAtEnd + 4;
      if (after >= rest.length || rest.slice(after).trim() === "") {
        yamlEnd = posAtEnd;
        bodyOffset = rest.length; // no body
      } else {
        yamlEnd = posAtEnd;
        bodyOffset = after;
      }
    } else {
      // No closing ---, return as-is with error behavior matching Rust (returns error)
      // but for resilience in mobile, treat as no frontmatter
      return { frontmatter: emptyResult, body: content };
    }
  }

  const yamlStr = rest.slice(0, yamlEnd).trim();
  const body = bodyOffset < rest.length ? rest.slice(bodyOffset) : "";

  if (yamlStr.length === 0) {
    return { frontmatter: emptyResult, body };
  }

  // Manual YAML parsing to match Rust yaml_rust2 behavior
  const frontmatter = parseYamlFields(yamlStr);

  return { frontmatter, body };
}

/**
 * Parse YAML key-value pairs from frontmatter string.
 * Handles: quoted strings, booleans, arrays (both inline [] and multi-line - "item").
 */
function parseYamlFields(yamlStr: string): FrontmatterData {
  const result: FrontmatterData = {
    id: null,
    title: null,
    tags: [],
    created_at: null,
    updated_at: null,
    is_pinned: false,
  };

  const lines = yamlStr.split("\n");
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];
    const colonIdx = line.indexOf(":");
    if (colonIdx === -1) {
      i++;
      continue;
    }

    const key = line.slice(0, colonIdx).trim();
    const rawValue = line.slice(colonIdx + 1).trim();

    switch (key) {
      case "id":
        result.id = parseYamlString(rawValue);
        break;
      case "title":
        result.title = parseYamlString(rawValue);
        break;
      case "created_at":
        result.created_at = parseYamlString(rawValue);
        break;
      case "updated_at":
        result.updated_at = parseYamlString(rawValue);
        break;
      case "is_pinned":
        result.is_pinned = rawValue === "true";
        break;
      case "tags": {
        // Could be inline `tags: []` or `tags:` followed by `  - "item"` lines
        if (rawValue === "[]" || rawValue === "") {
          // Check if next lines are array items
          if (rawValue === "") {
            const items: string[] = [];
            while (i + 1 < lines.length) {
              const nextLine = lines[i + 1];
              const trimmed = nextLine.trim();
              if (trimmed.startsWith("- ")) {
                items.push(parseYamlString(trimmed.slice(2).trim()));
                i++;
              } else {
                break;
              }
            }
            result.tags = items;
          } else {
            result.tags = [];
          }
        } else if (rawValue.startsWith("[")) {
          // Inline array: [tag1, tag2] or ["tag1", "tag2"]
          const inner = rawValue.slice(1, -1).trim();
          if (inner === "") {
            result.tags = [];
          } else {
            result.tags = inner.split(",").map((s) => parseYamlString(s.trim()));
          }
        } else {
          // Single value on same line (unlikely but handle it)
          result.tags = [parseYamlString(rawValue)];
        }
        break;
      }
      default:
        break;
    }

    i++;
  }

  return result;
}

/**
 * Parse a YAML string value, handling quoted and unquoted forms.
 * Removes surrounding quotes and unescapes \" sequences.
 */
function parseYamlString(value: string): string {
  if (value.length === 0) {
    return "";
  }

  // Double-quoted string
  if (value.startsWith('"') && value.endsWith('"')) {
    return value
      .slice(1, -1)
      .replace(/\\"/g, '"')
      .replace(/\\\\/g, "\\");
  }

  // Single-quoted string
  if (value.startsWith("'") && value.endsWith("'")) {
    return value.slice(1, -1).replace(/''/g, "'");
  }

  // Unquoted
  return value;
}
